src/traefik.py

```diff
@@ -0,0 +1,574 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Traefik workload interface."""
+import contextlib
+import enum
+import logging
+import re
+import socket
+from pathlib import Path
+from string import Template
+from typing import Any, Dict, List, Optional, cast
+
+import yaml
+from lightkube.models.core_v1 import ServicePort
+from ops import Container
+from ops.pebble import LayerDict, PathError
+from utils import is_hostname
+
+logger = logging.getLogger(__name__)
+DYNAMIC_CONFIG_DIR = "/opt/traefik/juju"
+STATIC_CONFIG_DIR = "/etc/traefik"
+STATIC_CONFIG_PATH = f"{STATIC_CONFIG_DIR}/traefik.yaml"
+DYNAMIC_CERTS_PATH = f"{DYNAMIC_CONFIG_DIR}/certificates.yaml"
+DYNAMIC_TRACING_PATH = f"{DYNAMIC_CONFIG_DIR}/tracing.yaml"
+SERVER_CERT_PATH = f"{DYNAMIC_CONFIG_DIR}/server.cert"
+SERVER_KEY_PATH = f"{DYNAMIC_CONFIG_DIR}/server.key"
+CA_CERTS_PATH = "/usr/local/share/ca-certificates"
+CA_CERT_PATH = f"{CA_CERTS_PATH}/traefik-ca.crt"
+RECV_CA_TEMPLATE = Template(f"{CA_CERTS_PATH}/receive-ca-cert-$rel_id-ca.crt")
+BIN_PATH = "/usr/bin/traefik"
+
+
+class RoutingMode(enum.Enum):
+    """Routing mode."""
+
+    path = "path"
+    subdomain = "subdomain"
+
+
+class TraefikError(RuntimeError):
+    """Base class for errors raised by this module."""
+
+
+class ContainerNotReadyError(TraefikError):
+    """Raised if the caller incorrectly assumes that the traefik container is ready."""
+
+
+class Traefik:
+    """Traefik workload representation."""
+
+    diagnostics_port = 8082  # Prometheus metrics, healthcheck/ping
+    _port = 80
+    _tls_port = 443
+    log_path = "/var/log/traefik.log"
+    _layer_name = "traefik"
+
+    def __init__(
+        self,
+        service_name: str,
+        container: Container,
+        routing_mode: RoutingMode,
+        tls_enabled: bool,
+        tcp_entrypoints: Dict[str, int],
+    ):
+        self._service_name = service_name
+        self._container = container
+        self._tcp_entrypoints = tcp_entrypoints
+        self._routing_mode = routing_mode
+        self._tls_enabled = tls_enabled
+
+    @property
+    def _hostname(self) -> str:
+        return socket.getfqdn()
```

Comment:
I'm on the fence with having this property.
- In ingress requirers it makes more sense to have `_internal_url` but for traefik we don't need it probable.
- Calling this `hostname` is confusing because it's actually "internal hostname"
- Testing is a bit more fragile because what do we mock? _hostname or getfqdn?
- `socket.getfqdn()` is just as easy to type as `self._hostname`

I guess I'm in favor of using getfqdn directly.